### PR TITLE
[torch.optim/C++] Add Adagrad state initialization

### LINKED ISSUE #87706: C++ Adagrad optimizer doesn't initialize parameter state
### 🐛 Describe the bug

optim/adagrad.cpp lacks code to initialize parameter state.

Running this code:

```
 auto x=torch::tensor({0.5, 2.0, 4.0}, torch::requires_grad());
 auto y=torch::tensor({1.0, 2.0, 3.0});
 auto o=torch::optim::Adagrad(std::vector<torch::Tensor>{}, torch::optim::AdagradOptions(0.1));
 auto& p=o.param_groups();
 p[0].params().push_back(x);
 auto l=torch::nn::functional::mse_loss(x,y);
 l.backward();
 o.step();
```
will throw:
_'state_[c10::guts::to_string(p.unsafeGetTensorImpl())] != nullptr INTERNAL ASSERT FAILED at "../torch/csrc/api/src/optim/adagrad.cpp":80, please report a bug to PyTorch. state found NULL for the Tensor_ 

the other optimizers have code that searches for the state and initializes if not found, e.g. from sgd.cpp:

```
        auto param_state = state_.find(c10::guts::to_string(p.unsafeGetTensorImpl()));
        if(param_state == state_.end()) {
          buf = torch::clone(d_p).detach();
          auto state = std::make_unique<SGDParamState>();
          state->momentum_buffer(buf);
          state_[c10::guts::to_string(p.unsafeGetTensorImpl())] = std::move(state);
        } else {
          buf = static_cast<SGDParamState&>(*param_state->second).momentum_buffer();
          buf.mul_(momentum).add_(d_p, 1 - dampening);
        }
```
adagrad.cpp doesn't contain this kind of code.

### Versions

This error in all versions since PR  #29335

### PR DESCRIPTION
**Pull request by defoishugo**
Fixes #87706 by adding state initialization to the Adagrad optimizer.

cc @albanD @jbschlosser @soumith @iramazanli @vincentqb

### CODE REVIEW

File: torch/csrc/api/src/optim/adagrad.cpp [hunk 1]
```diff
@@ -76,13 +76,23 @@ Tensor Adagrad::step(LossClosure closure) {
...
       auto grad = p.grad();
-      TORCH_INTERNAL_ASSERT(
-          state_[c10::guts::to_string(p.unsafeGetTensorImpl())] != nullptr,
-          "state found NULL for the Tensor ",
-          p);
+
+      auto param_state =
+          state_.find(c10::guts::to_string(p.unsafeGetTensorImpl()));
+      auto& options = static_cast<AdagradOptions&>(group.options());
+      auto initial_accumulator_value = options.initial_accumulator_value();
```
Comment (janeyx99) [inline torch/csrc/api/src/optim/adagrad.cpp:83]: Does this handle complex?

**Review by janeyx99 — COMMENTED** [review]
Thanks for taking this on! Things look p good, I have 2 questions:
1. Does the implementation handle complex values?
2. Could you add test cases showing that the state found NULL error is no longer thrown?

**Comment by defoishugo** [comment]
@janeyx99 Thank you for your review.

> 1. Does the implementation handle complex values?

Despite the fact that the Python implementation supports complex values, the proposed C++ implementation does not. 

Unless I'm mistaken, other optimizers like `Adam` don't support complex tensors in the C++ implementation either, unlike the Python implementation. 

To be honest, I thought Porch didn't directly support complex tensors in C++ in the same way as real tensors. However, if you want us to add support for complex tensors, here's an idea (which I can't really test):

      if (p.is_complex()) {
        auto a = torch::scalar_tensor(initial_accumulator_value);
        auto c = torch::complex(a, a);
        auto filled_tensor = torch::empty(p.sizes(), MemoryFormat::Preserve);
        filled_tensor.fill_(c);
        state->sum(filled_tensor);
      }
      else {
        state->sum(torch::full_like(p, initial_accumulator_value, MemoryFormat::Preserve));
      }

Of course, code style would be changed for the PR.

> 2. Could you add test cases showing that the state found NULL error is no longer thrown?

As far as testing is concerned, I'll be doing a commit soon.

I'd also like to point out that a code was originally designed to initialize the state in `adagrad.h`. It seems that not only does this code not work (I don't know why), but more importantly, the other optimizers initialize the state in the `Adagrad::step` function. So I'm going to add a commit to delete the associated code.

**Comment by defoishugo** [comment]
@janeyx99 PR ready for review.

File: test/cpp/api/optim.cpp [hunk 2]
```diff
@@ -524,3 +524,14 @@ TEST(OptimTest, CheckLRChange_StepLR_Adam) {
 
   check_lr_change(optimizer, step_lr_scheduler, expected_epoch_lrs);
 }
+
+TEST(OptimTest, CheckStateInit_Adagrad) {
+  auto x = torch::tensor({0.5, 2.0, 4.0}, torch::requires_grad());
+  auto y = torch::tensor({1.0, 2.0, 3.0});
+  auto o = torch::optim::Adagrad(std::vector<torch::Tensor>{}, torch::optim::AdagradOptions(0.1));
```
Comment (janeyx99) [inline test/cpp/api/optim.cpp:531]: Why not include x in the std::vector here right off the bat?

**Review by janeyx99 — COMMENTED** [review]
Looks mostly good!

**Comment by janeyx99** [comment]
CI failures are legitimate, looks like an optim cpp test needs to be updated

**Comment by janeyx99** [comment]
If you overcome the merge conflicts, I can run another round of CI on it.

**Comment by defoishugo** [comment]
@janeyx99 Merge conflicts are resolved.

**Comment by defoishugo** [comment]
@janeyx99 Could you run another round of CI, please?

**Comment by defoishugo** [comment]
@janeyx99 Could you run another round of CI, please? One of the file which got committed earlier was containing a leak from another branch.

**Pull request closed without merging**